Sources/Private/CoreAnimation/Layers/ShapeLayer.swift

```diff
@@ -210,25 +210,21 @@ extension CALayer {
       .filter { !$0.hidden }
       .grouped(by: { $0 is Group })
 
-    // If this shape doesn't have any groups but just has top-level shape items,
-    // we can create a placeholder group with those items. (Otherwise the shape items
-    // would be silently ignored, since we expect all shape layers to have a top-level group).
-    if groupItems.isEmpty, parentGroup == nil {
-      groupItems = [Group(items: otherItems, name: "")]
+    // Handle the top-level `shapeLayer.items` array. This is typically just a single `Group`,
+    // but in practice can be any combination of items. The implementation expects all path-drawing
+    // shape items to be managed by a `GroupLayer`, so if there's a top-level path item we
+    // have to create a placeholder group.
+    if parentGroup == nil, otherItems.contains(where: { $0.drawsCGPath }) {
+      groupItems = [Group(items: items, name: "")]
       otherItems = []
     }
 
-    // `ShapeItem`s either draw a path, or modify how a path is rendered.
-    //  - If this group doesn't have any items that draw a path, then its
-    //    items are applied to all of this group's children.
-    let inheritedItemsForChildGroups: [ShapeItemLayer.Item]
-    if !otherItems.contains(where: { $0.drawsCGPath }) {
-      inheritedItemsForChildGroups = otherItems.map {
-        ShapeItemLayer.Item(item: $0, groupPath: parentGroupPath)
-      }
-    } else {
-      inheritedItemsForChildGroups = []
-    }
+    // Any child items that wouldn't be included in a valid shape render group
+    // need to be applied to child groups (otherwise they'd be silently ignored).
+    let inheritedItemsForChildGroups = otherItems
+      .map { ShapeItemLayer.Item(item: $0, groupPath: parentGroupPath) }
+      .shapeRenderGroups
+      .unusedItems
```

Comment:
Both of these changes are a superset of the previous logic. Always nice when we can just further-generalize existing logic rather than adding more special cases.